### DISCUSSION TITLE
Maccy 2.0.0 beta

### DISCUSSION THREAD

**Discussion by p0deje** [p0deje/Maccy#818] [Announcements]
### Overview

This is a discussion for beta releases of the Maccy 2.x version branch. The latest release is [2.0.0.beta.14](https://github.com/p0deje/Maccy/releases/tag/2.0.0.beta.14). This version can be run side-by-side with Maccy 1.0, so you can test it easily. It would be great if you could share the feedback on:

- How the new UI feels like.
- Performance issues.
- Memory utilization concerns.
- Dvorak or non-US keyboard layout issues.
- Any other issues not listed in **Known issues**.

### Highlights of 2.0

- The user interface has been rewritten from AppKit + NSMenu to SwiftUI + NSPanel.
- Storage has been rewritten from Core Data to SwiftData.
- Requires macOS 14 Sonoma or newer.
- Backward compatibility with Maccy 1.x. You can downgrade to 1.0 or newer if you don't like it or encounter bugs.
- No backward compatibility with Maccy０.x.
- New setting to pop up at the last position Maccy was opened at (#829).
- New search highlight matches mode with a yellow color similar to Safari (#830).
- ~~Maximum history size increased from 999 to 9999. _(This is being tested at the moment and might change before the final release)._~~ This was reverted due to multiple performance issues and is postponed until 2.1 or 2.2.

### Differences from 1.0

- Appearance preference **Number of items** takes no effect. Instead, vertically resize the main panel window according to your needs.
- Appearance preference **Title length** takes no effect. Instead, horizontally resize the main panel window according to your needs.
- Advanced preference **Avoid taking application focus** takes no effect. The focus will always remain on the active application, similar to how Spotlight works.

### Known Issues in beta

- [x] ~~If there are no copies, the window size is still big with empty vertical space.~~ Fixed in `beta.6`
- [x] ~~History size is not enforced and can grow beyond 999. This is done on purpose to test the new storage and see its limitations.~~ Fixed in `beta.7`
- [x] ~~Japanese language input in search doesn't work correctly. This seems to be a bug in SwiftUI (https://medium.com/@hajimenakamura/japanese-input-of-textfield-of-swiftui-does-not-working-70244684821d), since Korean/Chinese work correctly.~~ Fixed in `beta.7`
- [x] ~~(Un)pinning item doesn't immediately scroll to the item. It does not seem to be a big deal and is quite hard to fix, so likely it will not be fixed in the final release.~~ Fixed in `beta.10`
- [x] ~~Clipboard items access in the Shortcuts application misses files and images.~~ Fixed in `beta.9`
- [x] ~~Fast scrolling through a long history causes the scroll to bounce.~~ Fixed in `beta.4`
- [x] ~~Custom paste shortcuts are not working.~~ Fixed in `beta.5`
- [x] ~~Dvorak keyboard layout should work, but it wasn't extensively tested. Notably, pasting in the "Dvorak - QWERTY ⌘" keyboard layout might be broken.~~ Fixed in `beta.5` 
- [x] ~~Changing the pin letter or title requires restarting to take effect.~~ Fixed in `beta.3`
- [x] ~~Crashes on macOS Sequoia.~~ Fixed in `beta.2`
- [x] ~~A small window is shown behind the Maccy popup when clicking on a menu bar icon.~~ Fixed in `beta.8`
- [x] ~~Popup at screen center doesn't work properly with multiple screens.~~ Fixed in `beta.8`
- [x] ~~Pasting doesn't work when Maccy is opened by clicking on the menu icon.~~ Fixed in `beta.12`

**Comment by chungweileong94** [comment]
Been using the v2 for the past few hours, pretty solid to me, haven't got any problem so far. I didn't notice much about UI changes, but I like the new scrolling experience, way better compared to v1. In terms of memory usage, it's about the same as v1.

**Reply by p0deje** [reply]
> I didn't notice much about UI changes, but I like the new scrolling experience, way better compared to v1

That was my goal, have the similar UI look and feel backed by a different technology.

**Comment by reslear** [comment]
@p0deje 

1. Excellent UI, so modern, I think it was a great decision.
2. The search works perfectly.
3. Insertion/Copying without problems.
4. Memory ~150mb.
5. CPU in the background 0.0%!!! but under scroll load it can be <50%.
6. Размер приложения примерно тот-же ~3mb.

Спасибо за такой приятный update.

~~Из наблюдений:~~
- [x] ~~У меня 999 пунктом ощутимо подлагивает при скроле требует какой то оптимизации.~~
- [x] ~~Когда первый раз появляется окошко есть задержка, но потом когда уже скролишь оно уже появлется очень быстро без задержки (не знаю баг или фича) но можно увеличить задержку на пару секунд что бы оно не прыгало (не создавало визуальный шум) туда-сюда при скролинге.~~
~~https://github.com/user-attachments/assets/3eb6120a-6143-4660-9603-c35f01d0e264~~
___
**fixed in beta 12**

**Reply by p0deje** [reply]
> У меня 999 пунктом ощутимо подлагивает при скроле требует какой то оптимизации.

Можете записать видео? Сейчас есть проблема, что при скролле курсор может подсветить отдельную запись и остановить скролл.

> Когда первый раз появляется окошко есть задержка, но потом когда уже скролишь оно уже появлется очень быстро без задержки (не знаю баг или фича) но можно увеличить задержку на пару секунд что бы оно не прыгало (не создавало визуальный шум) туда-сюда при скролинге.

Так должно быть и в Maccy 1.0, это не новая фича.

**Reply by p0deje** [reply]
Попробуйте beta.4, скроллинг должен работать лучше.

**Reply by reslear** [reply]
@p0deje в beta 12 все проблемы выше - решены, Amazing!

**Comment by p0deje** [comment]
I've just released beta 2 which fixes Sequoia crashes, please try https://github.com/p0deje/Maccy/releases/tag/2.0.0.beta.2.

**Comment by p0deje** [comment]
I've just released beta 3 with small improvements and fixes, please see https://github.com/p0deje/Maccy/releases/tag/2.0.0.beta.3.

**Comment by p0deje** [comment]
I've just released beta 4 with better performance and small fixes, please see https://github.com/p0deje/Maccy/releases/tag/2.0.0.beta.4.

**Comment by p0deje** [comment]
I've just released beta 5 with better performance and small fixes, please see https://github.com/p0deje/Maccy/releases/tag/2.0.0.beta.5.

**Reply by F1248** [reply]
It seems that the `Maccy.app.zip` asset is missing from beta 5.

**Reply by p0deje** [reply]
My bad, upload failed and I haven't noticed. Please see again.

**Comment by p0deje** [comment]
I've just released beta 6 with improvements related to a popup height , please see https://github.com/p0deje/Maccy/releases/tag/2.0.0.beta.6.

**Comment by p0deje** [comment]
I've just released beta 7 with improvements related to history size and Japanese language, please see https://github.com/p0deje/Maccy/releases/tag/2.0.0.beta.7.

**Comment by p0deje** [comment]
New day, new beta 8 with fixes related to popup at screen center location.
https://github.com/p0deje/Maccy/releases/tag/2.0.0.beta.8.

**Comment by p0deje** [comment]
I've released beta 9 with mulitple fixes, please see https://github.com/p0deje/Maccy/releases/tag/2.0.0.beta.9.

**Comment by jaspermayone** [comment]
Am on the latest beta, is there a way to disable the maccy notification?

**Reply by p0deje** [reply]
Just like with 1.0, you have to disable in System Settings / Notifications.

**Comment by p0deje** [comment]
I've just released beta 10 which adds a new option to remember last position where Maccy was opened and addresses all known issues in Maccy 2.0 beta - https://github.com/p0deje/Maccy/releases/tag/2.0.0.beta.10 (credits to @weisJ)

**Comment by billchase2** [comment]
Are there plans to add a deduplication feature to Maccy 2.0? I've found on my older iMac that it doesn't always copy text when I cmd+c. So I've gotten into the habit of hitting the keyboard shortcut multiple times in a row to make sure it's copied. However, each time I do this, it saves the copied text to Maccy multiple times.

**Reply by p0deje** [reply]
Maccy deduplicates items, but sometimes when you copy rich text, every new copy gets new data (e.g. in Word) so it looks like duplicate.

**Reply by billchase2** [reply]
Ah, that makes sense. Thanks for the info!

**Comment by billchase2** [comment]
One more question: Are there plans to add syncing between devices?

**Reply by p0deje** [reply]
Yes, after 2.0 release, I would like to see how hard is it to add iCloud sync support in the new SwiftData storage stack.

**Reply by billchase2** [reply]
Excellent. Thank you!

**Comment by p0deje** [comment]
Please see the new beta 12 which fixes issues with pasting and glitchy UI - https://github.com/p0deje/Maccy/releases/tag/2.0.0.beta.12

**Reply by z0ffy** [reply]
When will the official version be released?

**Reply by p0deje** [reply]
As soon as I'm happy about its quality.

**Comment by z0ffy** [comment]
![image](https://github.com/user-attachments/assets/b80dea1a-9809-4cd1-bab1-a976844f0422)
Is the overlap of the search box a BUG?

**Reply by p0deje** [reply]
What OS version are you using?

**Reply by z0ffy** [reply]
15.0 Beta(24A5327a)

**Reply by p0deje** [reply]
I don't know if that's the issue on Sequoia or just a beta. I will look closer once I upgrade to a public beta.

**Reply by z0ffy** [reply]
> I don't know if that's the issue on Sequoia or just a beta. I will look closer once I upgrade to a public beta.

ok，thanks

**Reply by p0deje** [reply]
Fixed in beta 13

**Reply by z0ffy** [reply]
> Fixed in beta 13

You are so awesome.

**Comment by p0deje** [comment]
Please see the new beta 13 which fixes issues with pasting fuzzy search and adds a new search highlight matches mode with a yellow color similar to Safari. - https://github.com/p0deje/Maccy/releases/tag/2.0.0.beta.13

**Comment by z0ffy** [comment]
https://github.com/user-attachments/assets/f7b62af7-51c5-4edd-8392-4724b745669a
@p0deje Please watch this video. There will be a jumping situation when scrolling. It would be great if it could be repaired.

**Reply by chungweileong94** [reply]
I think it's because the mouse cursor keeps taking away the focus, while we navigate with keyboard.

**Reply by p0deje** [reply]
Is this an issue? I don’t personally care if it’s jumping and it clearly indicates that the end was reached.

**Comment by p0deje** [comment]
Please see the new beta 14 which fixes a couple of issues with pinned items (https://github.com/p0deje/Maccy/releases/tag/2.0.0.beta.14)

**Comment by luckydonald** [comment]
**Avoid taking application focus** is a show stopper for apps which close the input mask on de-focus.
I have that turned on for a while, as some apps aren't playing nice, if the text field looses focus.

**Reply by p0deje** [reply]
Have you tested 2.0 - it should **NOT** steal focus anymore. Any issues with specific apps?

**Comment by luckydonald** [comment]
**2.0.0.beta.14**: Selecting <kbd>Preferences…</kbd> > <kbd>🚫 Ignore</kbd> > <kbd>Pasteboard Types</kbd> crashes the app.

<img width="644" alt="Screenshot 2024-09-08 at 23 56 26" src="https://github.com/user-attachments/assets/30b630ed-1e65-471f-8ed1-077bcaf71622">

**1.0** works fine there.
<img width="612" alt="Screenshot 2024-09-09 at 00 00 02" src="https://github.com/user-attachments/assets/6cbee2a7-f324-4275-8836-12af6f179638">

**Reply by p0deje** [reply]
Can you please provide crash logs and output of running `defaults read org.p0deje.Maccy` in Terminal?

**Reply by F1248** [reply]
> Can you please provide crash logs and output of running `defaults read org.p0deje.Maccy` in Terminal?

This error also occurs to me and is 100% reproducible.

[Maccy-2024-09-09-191131.ips.txt](https://github.com/user-attachments/files/16934111/Maccy-2024-09-09-191131.ips.txt)

```
user@Mac ~ % defaults read org.p0deje.Maccy
{
    "KeyboardShortcuts_delete" = "{\\"carbonKeyCode\\":51,\\"carbonModifiers\\":2048}";
    "KeyboardShortcuts_pin" = 0;
    "KeyboardShortcuts_popup" = "{\\"carbonModifiers\\":2304,\\"carbonKeyCode\\":46}";
    "LaunchAtLogin__hasMigrated" = 1;
    NSNavLastRootDirectory = "/Applications";
    NSNavPanelExpandedSizeForOpenMode = "{800, 448}";
    "NSStatusItem Preferred Position Item-0" = 5559;
    "NSStatusItem Visible Item-0" = 1;
    "NSStatusItem Visible Item-1" = 0;
    "NSWindow Frame NSNavPanelAutosaveName" = "320 339 800 448 0 0 1440 875 ";
    "NSWindow Frame SUStatusFrame" = "491 556 400 134 0 0 1440 875 ";
    "NSWindow Frame SUUpdateAlert" = "372 357 638 398 0 0 1440 875 ";
    "NSWindow Frame com.sindresorhus.Preferences.FrameAutosaveName" = "454 368 532 262 0 0 1440 875 ";
    "NSWindow Frame com.sindresorhus.Settings.FrameAutosaveName" = "661 470 532 460 0 0 1920 1055 ";
    SUAutomaticallyUpdate = 1;
    SUEnableAutomaticChecks = 0;
    SUHasLaunchedBefore = 1;
    SULastCheckTime = "2024-07-24 09:11:34 +0000";
    SUUpdateGroupIdentifier = 1726747288;
    WebKitDefaultFontSize = 13;
    WebKitJavaScriptEnabled = 0;
    WebKitPluginsEnabled = 0;
    WebKitStandardFont = "-apple-system-font";
    WebKitUserStyleSheetEnabledPreferenceKey = 1;
    WebKitUserStyleSheetLocationPreferenceKey = "/Applications/Maccy.app/Contents/Frameworks/Sparkle.framework/Resources/ReleaseNotesColorStyle.css";
    avoidTakingFocus = 0;
    clearSystemClipboard = 1;
    enabledPasteboardTypes =     (
        "public.tiff",
        "public.png",
        "public.rtf",
        "public.utf8-plain-text",
        "public.html",
        "public.file-url"
    );
    highlightMatch = color;
    historySize = 999;
    ignoreAllAppsExceptListed = 0;
    ignoreEvents = 0;
    ignoreOnlyNextEvent = 0;
    ignoredPasteboardTypes =     (
        "de.petermaurer.TransientPasteboardType",
        "Pasteboard generator type",
        "com.agilebits.onepassword",
        "com.typeit4me.clipping",
        "net.antelle.keeweb"
    );
    imageMaxHeight = 200;
    maxMenuItemLength = 100;
    maxMenuItems = 0;
    menuIcon = maccy;
    migrations =     {
        "2020-04-25-allow-custom-ignored-types" = 1;
        "2020-06-19-use-keyboardshortcuts" = 1;
        "2020-09-01-ignore-keeweb" = 1;
        "2021-02-20-allow-to-customize-supported-types" = 1;
        "2021-06-28-add-title-to-history-item" = 1;
        "2021-10-16-remove-dynamic-pasteboard-types" = 1;
        "2022-08-01-rename-suppress-clear-alert" = 1;
        "2022-11-14-add-html-rtf-to-supported-types" = 1;
        "2023-01-22-add-regexp-search-mode" = 1;
        "2024-07-01-version-2" = 1;
    };
    pasteByDefault = 0;
    pinTo = top;
    playSounds = 0;
    popupPosition = center;
    popupScreen = 1;
    previewDelay = 200;
    searchMode = mixed;
    showFooter = 1;
    showInStatusBar = 1;
    showRecentCopyInMenuBar = 0;
    showSearch = 1;
    showTitle = 0;
    sortBy = lastCopiedAt;
    suppressClearAlert = 1;
    windowSize = "[473,547]";
}
```

**Reply by p0deje** [reply]
@F1248 This seems to be an issue related to Xcode 16 beta (https://forums.swift.org/t/confusion-around-mainactor-isolation-vs-ie-nonisolated-view-scrollposition/72597/18). I've tried compiling on the latest Xcode 16 RC, can you check if [this version](https://github.com/user-attachments/files/16935714/Maccy.app.zip) behaves better?

**Reply by p0deje** [reply]
@luckydonald could you check ☝️ too?

**Reply by F1248** [reply]
> I've tried compiling on the latest Xcode 16 RC, can you check if [this version](https://github.com/user-attachments/files/16935714/Maccy.app.zip) behaves better?

@p0deje, this version works fine, thanks!

**Reply by luckydonald** [reply]
@p0deje Works!
Tho the striped look of the old menu helped the readability IMO
<img width="1050" alt="Screenshot 2024-09-11 at 12 33 47" src="https://github.com/user-attachments/assets/5736d102-0eee-4922-adef-c472a77cee4e">

**Comment by ydq** [comment]
When the Chinese input method is enabled, the shortcut key `⌘`+`,` cannot open the settings panel. I have to switch to the English input method to use the shortcut key to open the settings. I tested other software and found that the shortcut key can be used to open the settings panel when using the Chinese input method.

Maccy demo：

https://github.com/user-attachments/assets/c6413f55-82c2-4cd4-8c6d-f44d3243c0ef

vscode demo:

https://github.com/user-attachments/assets/1e7201ee-9f94-4902-94a2-df8b0c0d610f

**Reply by p0deje** [reply]
Fixed in beta 15.

**Comment by p0deje** [comment]
Please see beta 15 which fixes few issues https://github.com/p0deje/Maccy/releases/tag/2.0.0.beta.15

**Comment by p0deje** [comment]
https://github.com/p0deje/Maccy/pull/869 is open, I am about to merge it into the main branch and make a final 2.0 release during the weekend.

**Comment by p0deje** [comment]
2.0 is released, thanks everyone for participating in beta and reporting bugs, you are awesome!

**Comment by bgpntx** [comment]
Not sure if it's expected or not but it seems that big sized screenshots or pictures are limited in preview, see attached screen.
<img width="678" alt="Screenshot at Sep 17 12-11-13" src="https://github.com/user-attachments/assets/6967b6f6-94c4-4299-98a5-410c644142a5">
1 and 2 are quite big screenshots, while 3 is relatively small, 1058 × 726 pixels. Maccy Version 2.0.0 (47) .

**Reply by p0deje** [reply]
Can you share examples of these screenshots? There is a logic in Maccy to size down previews/thumbnails so maybe that's a culprit.

**Reply by bgpntx** [reply]
For example:

**This file does not fit:**
<img width="1303" alt="image" src="https://github.com/user-attachments/assets/d3329742-a22d-44ce-9505-fcf583a83204">


**This file fits:**
<img width="233" alt="image" src="https://github.com/user-attachments/assets/b5d2a0dc-5e4a-4513-bd4d-0a9a1f944ee5">

**This is Maccy window:**
<img width="659" alt="image" src="https://github.com/user-attachments/assets/0aa79c9d-4a34-4a75-a337-1476101a2d0d">

Thanks!

**Reply by p0deje** [reply]
If I copy your "Doesn't fit" image from the browser, it shows just fine, so maybe that's because it's coming from Lightshot?
<img width="1534" alt="Screenshot 2024-09-17 at 08 14 31" src="https://github.com/user-attachments/assets/dde1c5b8-c02d-4b1c-a023-f2188a8b2632">

**Reply by bgpntx** [reply]
It appears so, that's weird... Okay, case closed :)

**Comment by bmmmm** [comment]
The release (2.2.0) is a very nice one again! 👍 
I am curious if there is another beta testing channel available? Did I overlook any information, or what is the current status of beta testing? 
Wishing everyone a pleasant weekend. 😊

**Reply by p0deje** [reply]
There was beta testing for 2.0, but now it's released and there are no more extra beta testing happening.

**Reply by bmmmm** [reply]
I have now switched back to the stable installation again. Thank you for the clarification 😊
